staking/staking_hub.go

```diff
@@ -0,0 +1,396 @@
+	"github.com/hashicorp/go-hclog"
+)
+
+type StakingEventType string
+
+var (
+	StakingEvent   StakingEventType = "staking"
+	UnstakingEvent StakingEventType = "unstaking"
+	UnknownEvent   StakingEventType = "unknown"
+)
+
+var (
+	defaultFileName = "staking-map.json"
+)
+
+// StakingHub acts as a hub (manager) for staked account balances
+type StakingHub struct {
+	// Address -> Stake
+	StakingMap map[types.Address]*big.Int
+
+	// Specifies the working directory for the Polygon SDK (location for writeback)
+	WorkingDirectory string
+
+	// Write-back period (in s) for backup staking data
+	WritebackPeriod time.Duration
+
+	// Logger for logging errors and information
+	Logger hclog.Logger
+
+	// Event queue defines staking / unstaking events
+	// which are read by modules that do final transaction sealing
+	EventQueue []PendingEvent
+
+	// Event list mutex
+	EventQueueMutex sync.Mutex
+
+	// Staking map mutex
+	StakingMutex sync.Mutex
+
+	// Close channel
+	CloseCh chan struct {
+	}
+}
+
+var stakingHubInstance StakingHub
+var once sync.Once
+
+// GetStakingHub initializes the stakingHubInstance singleton
+func GetStakingHub() *StakingHub {
+	once.Do(func() {
+		stakingHubInstance = StakingHub{
+			StakingMap:      make(map[types.Address]*big.Int),
+			EventQueue:      make([]PendingEvent, 0),
+			CloseCh:         make(chan struct{}),
+			WritebackPeriod: 60 * time.Second,
+		}
+	})
+
+	return &stakingHubInstance
+}
+
+// SetWorkingDirectory sets the writeback directory for the staking map
+func (sh *StakingHub) SetWorkingDirectory(directory string) {
+	sh.WorkingDirectory = directory
+
+	err := sh.readFromDisk()
+	if err != nil {
+		// Log as warning because reading this map is not crucial
+		sh.log(err.Error(), logWarning)
+	}
+
+	go sh.saveToDisk()
+}
+
+// LogType defines possible log types for the logger
+type LogType string
+
+var (
+	logInfo    LogType = "info"
+	logError   LogType = "error"
+	logWarning LogType = "warning"
+)
+
+// log logs the output to the console if the logger is set
+func (sh *StakingHub) log(output string, logType LogType) {
+	if sh.Logger != nil {
+		switch logType {
+		case logInfo:
+			sh.Logger.Info(output)
+		case logError:
+			sh.Logger.Error(output)
+		case logWarning:
+			sh.Logger.Warn(output)
+		}
+	}
+}
+
+// SetLogger sets the StakingHub logger
+func (sh *StakingHub) SetLogger(logger hclog.Logger) {
+	sh.Logger = logger.Named("staking-hub")
+}
+
+// CloseStakingHub stops the writeback process
+func (sh *StakingHub) CloseStakingHub() {
+	sh.StakingMutex.Lock()
+	defer sh.StakingMutex.Unlock()
+
+	// Alert the closing channel
+	sh.CloseCh <- struct{}{}
+
+	close(sh.CloseCh)
+}
+
+// PendingEvent contains useful information about a staking / unstaking event
+type PendingEvent struct {
+	BlockNumber int64
+	Address     types.Address
+	Value       *big.Int
+	EventType   StakingEventType
+}
+
+// Compare checks if the two events match
+func (pe *PendingEvent) Compare(event PendingEvent) bool {
```

Comment:
`Compare` should usually return `-1`, `0`, `1` since it is checking how 2 values compare to each other.
`Equal` should usually return true or false.

I think you should rename this method to `Equal`

Reply:
Completely agree 👍 

Fixed 👍 

ebc42eb